azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+package azkaban.executor.container.watch;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import azkaban.Constants.ContainerizedDispatchManagerProperties;
+import azkaban.executor.AlerterHolder;
+import azkaban.executor.ExecutableFlow;
+import azkaban.executor.ExecutionControllerUtils;
+import azkaban.executor.ExecutorLoader;
+import azkaban.executor.ExecutorManagerException;
+import azkaban.executor.Status;
+import azkaban.executor.container.ContainerizedImpl;
+import azkaban.utils.Props;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheStats;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides callback implementations of {@link AzPodStatusListener} for
+ * (1) Updating WebServer and database state of flows based on Pod status
+ * (2) Driving pod lifecycle actions, such as deleting flow pods in a final state.
+ */
+public class FlowStatusUpdatingListener implements AzPodStatusListener {
+
+  private static final Logger logger = LoggerFactory.getLogger(FlowStatusUpdatingListener.class);
+  public static final int EVENT_CACHE_STATS_FREQUENCY = 100;
+  public static final int DEFAULT_EVENT_CACHE_MAX_ENTRIES = 4096;
+
+  private final ContainerizedImpl containerizedImpl;
+  private final ExecutorLoader executorLoader;
+  private final AlerterHolder alerterHolder;
+  private final Cache<String, AzPodStatusMetadata> podStatusCache;
+
+  // Note about the cache size.
+  // Each incoming event is expected to be less than 5KB in size and the maximum cache size will be
+  // about 5KB * maxCacheEntries.
+  private final int maxCacheEntries;
+
+  // Convenience member for referring to the Cache through ConcurrentMap interface.
+  private final ConcurrentMap<String, AzPodStatusMetadata> podStatusMap;
+  private final AtomicLong flowContainerEventCount = new AtomicLong(0);
+
+  @Inject
+  public FlowStatusUpdatingListener(Props azkProps,
+      ContainerizedImpl containerizedImpl,
+      ExecutorLoader executorLoader,
+      AlerterHolder alerterHolder) {
+    requireNonNull(azkProps, "azkaban properties must not be null");
+    requireNonNull(containerizedImpl, "container implementation must not be null");
+    requireNonNull(executorLoader, "executor loader must not be null");
+    requireNonNull(alerterHolder, "alerter holder must not be null");
+    this.containerizedImpl = containerizedImpl;
+    this.executorLoader = executorLoader;
+    this.alerterHolder = alerterHolder;
+
+    maxCacheEntries =
+        azkProps.getInt(ContainerizedDispatchManagerProperties.KUBERNETES_WATCH_DRIVER_THREAD_POOL_SIZE,
```

Comment:
The property and default value names are confusing. One is for thread pool and other is for cache max entries?